magma/passes/when.py

```diff
@@ -0,0 +1,38 @@
+from magma.passes.passes import DefinitionPass, pass_lambda
+from magma.primitives.when import WhenBuilder
+
+
+class WhenPass(DefinitionPass):
+    def __call__(self, definition):
+        with definition.open():
+            for builder in definition._context_._builders:
+                if isinstance(builder, WhenBuilder):
+                    self.process_when_builder(builder, definition)
+
+
+class InferLatches(WhenPass):
+    def process_when_builder(self, builder, defn):
+        builder.infer_latches()
+
+
+class EmitWhenAsserts(WhenPass):
+    def process_when_builder(self, builder, defn):
+        builder.emit_when_asserts()
+
+
+class FinalizeWhens(WhenPass):
+    def process_when_builder(self, builder, defn):
+        assert not builder._finalized
+        defn._context_._placer.place(builder.finalize())
+
+
+infer_latches = pass_lambda(InferLatches)
+do_emit_when_asserts = pass_lambda(EmitWhenAsserts)
```

Comment:
i think we should name these as `xxx_pass`. then we can get rid of `do_emit_...` (assuming the do_ is to avoid confusion with the emit_when_asserts bool option)